draft-ietf-tls-esni.md

```diff
@@ -275,25 +281,30 @@ or if names can be added or removed from the anonymity set during the lifetime
 of a particular ECH configuration.
 
 public_name
-: The non-empty name of the client-facing server, i.e., the entity trusted to
-update the ECH configuration. This is used to correct misconfigured clients, as
-described in {{handle-server-response}}. This value MUST NOT begin or end with
-an ASCII dot and MUST be parsable as a dot-separated sequence of LDH labels, as
-defined in {{!RFC5890, Section 2.3.1}}. Clients MUST ignore any `ECHConfig`
-structure whose `public_name` does not meet these criteria. Note that these
-criteria are incomplete; they incidentally rule out textual representations of
-IPv6 addresses (see {{!RFC3986, Section 3.2.2}}), but do not exclude IPv4
-addresses in standard dotted-decimal or other non-standard notations such as
-octal and hexadecimal (see {{RFC3986, Section 7.4}}). If `public_name` contains
-a literal IPv4 or IPv6 address, the client SHOULD ignore the `ECHConfig` to
-avoid sending a non-compliant "server_name" extension on the ClientHelloOuter
-(see {{!RFC6066, Section 3}}).
+: The non-empty DNS name of the client-facing server, i.e., the entity trusted
+to update the ECH configuration. This is used to correct misconfigured clients,
+as described in {{handle-server-response}}. This value MUST NOT begin or end
+with an ASCII dot and MUST be parsable as a dot-separated sequence of LDH
+labels, as defined in {{!RFC5890, Section 2.3.1}}. Clients MUST ignore any
+`ECHConfig` structure whose `public_name` does not meet these criteria. Note
+that these criteria allow IPv4 addresses in standard dotted-decimal or other
+non-standard notations such as octal and hexadecimal (see {{?RFC3986}}, Section
+7.4). Clients SHOULD ignore the `ECHConfig` if it contains an encoded IPv4
+address. (To determine if a public_name value is an IPv4 address, clients can
+invoke the IPv4 parser algorithm in {{WhatWGURLAddressParser}}. It returns a
+value when the input is an IPv4 address.) Failure to validate that a public_name
+is not an IPv4 address may cause connection failures to the client-facing
+server. However, it is not expected to induce security failures. Clients will
+fail closed if the client-facing server cannot produce a certificate bound
+to the reference identity based on the public_name.
```

Comment:
This assertion requires a bunch of assumptions:
* A server has published an IPv4 address in the public_name even though they're not supposed to, presumably because it's permitted by a future draft.
* public_name is delivered "securely".
* ICANN won't let anyone register a numeric TLD.
* Clients don't parse IPv4 strings into 4 octets, and then try to interpret that as a domain name, resulting in a memory safety violation.

I find these assumptions less than totally obvious, and that makes this text confusing.

Here's some alternative explanatory text.
```suggestion
value when the input is an IPv4 address.) Ignoring IPv4 addresses in `public_name`
allows their use to be defined in a future update to ECH without breaking
compatibility.
```

Reply:
> Ignoring IPv4 addresses in `public_name` allows their use to be defined in a future update to ECH without breaking compatibility.

I don't think this is true. If that's the goal, we'd need a MUST-level requirement. With a SHOULD, some TLS implementations will ignore it, while others will go ahead and use that ECHConfig and likely get confused.

Reply:
@bemasc clarifying question: by "ignoring IPv4 addresses in `public_name`," do you mean "ignore ECHConfigs that contain an IPv4 address," or "allow ECHConfigs that contain an IPv4 address (by ignoring the fact that the config contains one)"? (I'm parsing your suggestion as the latter, which seems counter to what this PR is trying to do.)

Reply:
@davidben Perhaps, but that's how I read the intention of the PR nonetheless.

@chris-wood The former.  I updated my comment to clarify.